Add network-adapters inventory manager.

ikmsk10 opened the pull request:
This patch brings the following changes:
* storage: improve pciids lookup
* storage: rename inventory manager interface.
* network: add network-adapters inventory manager.

ikmsk10 committed:
Add network-adapters inventory manager.

The appropriate service provides objects with the
com.Yadro.NetworkAdapter interface.

 Example:
```
busctl introspect com.yadro.NetworkAdapter
       /xyz/openbmc_project/inventory/system/network/adapter/eth0

NAME                                TYPE      SIGNATURE RESULT/VALUE
org.freedesktop.DBus.Introspectable interface -         -
.Introspect                         method    -         s
org.freedesktop.DBus.Peer           interface -         -
.GetMachineId                       method    -         s
.Ping                               method    -         -
org.freedesktop.DBus.Properties     interface -         -
.Get                                method    ss        v
.GetAll                             method    s         a{sv}
.Set                                method    ssv       -
.PropertiesChanged                  signal    sa{sv}as  -
xyz...Decorator.Asset               interface -         -
.BuildDate                          property  s         ""
.Manufacturer                       property  s         "Intel Corporation"
.Model                              property  s         "I210 Gigabit Network Connection"
.PartNumber                         property  s         ""
.SerialNumber                       property  s         ""
xyz.openbmc_project.Inventory.Item  interface -         -
.Present                            property  b         true
.PrettyName                         property  s         "eth0"
xyz...NetworkInterface              interface -         -
.MACAddress                         property  s         "a0:c5:f2:10:28:3c"
xyz...OperationalStatus             interface -         -
.Functional                         property  b         true

```

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

alatarum commented on com/yadro/NetworkAdapter/Manager.interface.yaml:
```diff
@@ -0,0 +1,6 @@
+description: >
+    Control Hardware Manager daemon
```
I'm now wonder if it was good idea to add that com.yadro.Storage.Manager interface as a separate category... Probably it should be com.yadro.HWManager.Manager to not duplicate same interfaces.

ikmsk10 replied:
Yes, I was thought too about it. But I don't desire to change the previous implementation `Storage` in the context 'Add support Network Adapter manager'.
So, we should decide, does it is required to refine Storage implementation. Please, note, that changes will impact on the IPMI implementation, the part where we parse incoming storage information from the host.

alatarum replied:
Right, this will affect two places - hwmanager and ipmi plug in. But you modify both anyway.

ikmsk10 replied:
I guess that will be separate changes and patches anyway.

alatarum replied:
ok

nest1ing replied:
I like the idea about share the same interface. And now is a good time to do it.
If you don't want to mix these changes you could make it as separate PRs.
Although I see nothing wrong to combine these changes into one PR but by the different commits.

So, this PR looks wrong as it increases the code duplication.

ikmsk10 replied:
Fixed

alatarum commented on src/dbus.hpp:
```diff
@@ -35,6 +35,12 @@ constexpr const char* busName = "com.yadro.Storage";
+namespace netadpmgr
+{
+constexpr const char* busName = "com.yadro.NetworkAdapter";
+constexpr const char* path = "/com/yadro/network/adapter";
```
I think, /com/yadro/network would be enough.

ikmsk10 replied:
I believe the network is a too generic name and it aggregates too wide a definition. We should care about that and I'd prefer to explicitly define `network-adapter`, not general `network`.

alatarum replied:
Then call it `network_adapter`. I don't like an idea to add unneeded hierarchy level.

ikmsk10 replied:
And what do you think about the case when in the future, for example, we will be managing the host `network-vlan`s? 
By your design, that will be `network_vlan`. But that both things are network origin.
I believe the second level of the hierarchy is a good point there

alatarum replied:
Ok.

alatarum commented on src/network/adapter.cpp:
```diff
@@ -0,0 +1,101 @@
+const std::string NetworkAdapter::inventoryPath =
+    dbus::inventory::pathBase + std::string("/system/network/adapter/");
+
+const std::pair<std::string, std::string>
```
Let's move this to common header

ikmsk10 replied:
Hmm, that has some sense. I agree.

ikmsk10 replied:
But I believe the PCIVendorLookup from `storage/inventory.cpp/hpp` should be removed.
And extra issue: what do you think about renaming `storage/inventory.cpp/hpp` to something `storage/drive.cpp/hpp`?
I guess `inventory` is a too general definition.

alatarum replied:
> But I believe the PCIVendorLookup from storage/inventory.cpp/hpp should be removed.

Right, we can use your implementation.

> And extra issue: what do you think about renaming storage/inventory.cpp/hpp to something storage/drive.cpp/hpp?

My intention was that lately there will be some additional functionality added, like control RAID configuration and manage volumes, so "inventory" here was to show functionality.

nest1ing replied:
>  Let's move this to common header

+1

ikmsk10 replied:
FIxed

alatarum commented on src/network/adapter.cpp:
```diff
@@ -0,0 +1,101 @@
+    // xyz.openbmc_project.Inventory.Item
+    prettyName(name);
+    present(true);
+    // xyz.openbmc_project.Inventory.NetworkInterface
```
I wonder if Phosphor Network Manager (or how is it called) not confused by this new interfaces?

ikmsk10 replied:
The phosphor-dbus-interfaces already have `xyz.openbmc_project.Inventory.Item.NetworkInterface` and the answer to your question: no. The Inventory doesn't relate anyway with `phosphor-network-manager`.

ikmsk10 replied:
Oh, my bad. I'm too wrong.
Looks again on the `phosphor-networkd`, I noted the `xyz.openbmc_project.Inventory.Item.NetworkInterface` used to watch on the mac-address changes(I guess in BMC context).
And that is too strange for me. I still thought the `xyz.openbmc_project.Inventory` purpose is providing an inventory of Host, not of BMC. I was a little bit confused.

alatarum replied:
Yes, I also noticed that xyz.openbmc_project.Inventory.Item.NetworkInterface used to manage BMC network, and that is why I ask.

ikmsk10 replied:
The `xyz.openbmc_project.Inventory.Item.NetworkInterface` in the current service doesn't impact the phoshpor-networkd because the `Sync mac address with phosphor-inventory-manager` was processed only when `phosphor-networkd`:`sync-mac` is turned on.

Totally I very disliked that `xyz.openbmc_project.Inventory.Item` namespace which uses to provide BMC-specific information, I'd strongly like to avoid that feature.

nest1ing commented on src/network/adapter.cpp:
```diff
@@ -0,0 +1,101 @@
+    {
+        std::string curVIDstr = line.substr(vidPrefixLen, wordLen);
+        if (vendorName.empty() &&
+            (curVIDstr.compare(vendorId.substr(hexPrefixLen)) == 0))
```
I dislike the `substr` using as a temporary value because it leads to an excess allocation for this temporary string.
You could use the other overload of `compare` with a position as an argument.
https://en.cppreference.com/w/cpp/string/basic_string/compare

ikmsk10 replied:
fixed

nest1ing commented on src/network/main.cpp:
```diff
@@ -0,0 +1,95 @@
+        return;
+    }
+
+    buffer << dataFile.rdbuf();
```
`buffer` looks excess. `nlohmann/json` can read the data right from `fstream` object.
https://github.com/nlohmann/json#tofrom-streams-eg-files-string-streams

ikmsk10 replied:
fixed

ikmsk10 committed:
storage: Improve pciids lookup

This patch brings enhancement of pciLookup to provide a feature to
search PCI-device model by its DID.

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>

ikmsk10 committed:
storage: Rename inventory manager interface.

The `com.yadro.Storage.Manager` was renamed to the
`com.yadro.Inventory.Manager` to provide common interface to managaring
an YADRO-specific inventory.

End-user-impact: None

Signed-off-by: Igor Kononenko <i.kononenko@yadro.com>
